src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,41 +183,153 @@ internal int EndRead(IAsyncResult asyncResult)
+            ValidateParameters(buffer, offset, count);
+
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
             {
-                throw new ArgumentNullException(nameof(asyncResult));
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "WriteAsync", "write"));
+            }
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
+            {
+                // If it's an empty message and the PAL doesn't support that,
+                // we're done.
+                _nestedWrite = 0;
+                return Task.CompletedTask;
             }
+            if (count <= _sslState.MaxDataSize)
+            {
+                //Single write
+                Task t = WriteSingleChunk(buffer, offset, count, cancellationToken);
+                if (t.IsCompleted)
+                {
+                    _nestedWrite = 0;
 
-            LazyAsyncResult lazyResult = asyncResult as LazyAsyncResult;
-            if (lazyResult == null)
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
+            }
+            else
             {
-                throw new ArgumentException(SR.Format(SR.net_io_async_result, asyncResult.GetType().FullName), nameof(asyncResult));
+                //Loop
+                return WriteAsyncChunked(buffer, offset, count, cancellationToken);
             }
 
-            if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+            async Task ExitWriteAsync(Task task)
             {
-                throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "EndWrite"));
+                try
+                {
+                    await task;
+                }
+                catch (Exception e)
+                {
+                    _sslState.FinishWrite();
+
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
+
+                    throw new IOException(SR.net_io_write, e);
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
+        }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
+        private async Task WaitForWriteIOSlot(Task lockTask, byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            await lockTask;
+            await WriteSingleChunk(buffer, offset, count, cancellationToken);
+        }
 
-            if (lazyResult.Result is Exception e)
+        internal Task WriteSingleChunk(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            // Request a write IO slot.
+            Task ioSlot = _sslState.CheckEnqueueWriteAsync();
+            if (!ioSlot.IsCompleted)
```

Comment:
Is it possible ioSlot could complete as faulted rather than successfully?  If so, I'd suggest making this IsCompletedSuccessfully rather than IsCompleted.  That won't hurt the fast case, and if there is a chance it could fail, it'll ensure the exception is propagated rather than dropped, as we'll call WaitForWriteIOSlot which will await it.

Reply:
It would only do so if someone was mixing both a sync renegotiation and an async write. Can change but iscompletedsucessfully .net desktop or core only?

Reply:
IsCompletedSuccessfully is currently only in .NET Core.  But then, so is all of this code, no?

Reply:
Honestly the interaction between corefx -> core -> AOT -> desktop confuses me, if it's core only that's cool :) I only really code in that these days anyway.